ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Mempool.hs

```diff
@@ -151,20 +175,96 @@ applyHelper apply
     injectTxs :: InPairs InjectTx xs
     injectTxs = InPairs.requiringBoth cfgs hardForkInjectTxs
 
+    applyCurrent
+      :: forall blk. SingleEraBlock blk
+      => Index xs                                            blk
+      -> WrapLedgerConfig                                    blk
+      -> Product GenTx (Ticked :.: LedgerState)              blk
+      -> (Except (HardForkApplyTxErr xs) :.: ApplyResult xs) blk
+    applyCurrent index cfg (Pair tx (Comp st)) =
+      Comp
+        $ withExcept (injectApplyTxErr index)
+        $ do
+          (st', vtx) <- apply (unwrapLedgerConfig cfg) slot tx st
+          pure ApplyResult {
+              arValidatedTx = injectValidatedGenTx index vtx
+            , arState       = st'
+            }
+
+-- | A private type used only to clarify the definition of 'applyHelper'
+data ApplyResult xs blk = ApplyResult {
+    arValidatedTx :: Validated (GenTx (HardForkBlock xs))
+  , arState       :: Ticked (LedgerState blk)
+  }
+
+reapplyHelper
```

Comment:
This is another place where I've gone with duplication, at least at first

Reply:
As in `InjectTxs.hs`:

I've pushed up some additional commits that remove this duplication by instead generalizing the definition. We need to discuss whether the additional abstraction is worthwhile. I'm feeling that it is -- but I just wrote, so I may be underestimating it.

There are also other options, such as introducing a new GADT as a parameter or introducing type classes etc instead of abstracting out the type-specific parts as HOF arguments.

Reply:
You were comfortable with the de-duplication in `InjectTxs.hs`. How about the deduplication approach taken in this module?

Reply:
Making use of the functions defined in `InjectTxs.hs`, I think it is reasonable to change it to what you did on this other side. I think this looks good too.